pkg/storegateway/series_refs.go

```diff
@@ -36,18 +49,52 @@ type seriesChunkRefsIterator interface {
 type seriesChunkRefsSet struct {
 	// series sorted by labels.
 	series []seriesChunkRefs
+
+	// releasable holds whether the series slice (but not its content) can be released to a memory pool.
+	releasable bool
 }
 
-func newSeriesChunkRefsSet(capacity int) seriesChunkRefsSet {
+// newSeriesChunkRefsSet creates a new seriesChunkRefsSet with the given capacity.
+// If releasable is true, then a subsequent call release() will put the internal
+// series slices to a memory pool for reusing.
+func newSeriesChunkRefsSet(capacity int, releasable bool) seriesChunkRefsSet {
+	var prealloc []seriesChunkRefs
+
+	// If it's releasable then we try to reuse a slice from the pool.
+	if releasable {
+		if reused := seriesChunkRefsSetPool.Get(); reused != nil {
+			prealloc = *(reused.(*[]seriesChunkRefs))
+		}
+	}
+
+	if prealloc == nil {
+		prealloc = make([]seriesChunkRefs, 0, capacity)
+	}
+
 	return seriesChunkRefsSet{
-		series: make([]seriesChunkRefs, 0, capacity),
+		series:     prealloc,
+		releasable: releasable,
 	}
 }
 
 func (b seriesChunkRefsSet) len() int {
 	return len(b.series)
 }
 
+// release the internal series slice to a memory pool. This function call has no effect
+// if seriesChunkRefsSet was created to be not releasable.
```

Comment:
I'd also mention that `release` is idempotent

Reply:
Unfortunately it's not. See the TODO below.